src/hotspot/share/opto/addnode.cpp

```diff
@@ -396,9 +396,179 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  if (mul.is_valid_with(rhs)) {
+    return mul;
+  }
+
+  // (4) Power-of-two addition pattern (e.g., lhs = (a << CON1) + (a << CON2))
+  // While multiplications can be potentially optimized to power-of-2 subtractions (e.g., a * 7 => (a << 3) - a),
+  // (x - y) + y => x is already handled by the Identity() methods. So, we don't need to check for that pattern here.
+  mul = find_power_of_two_addition_pattern(lhs, bt);
+  if (mul.is_valid_with(rhs)) {
+    return mul;
+  }
+
+  // We've tried everything.
+  return Multiplication::make_invalid();
+}
+
+// Try to match `n = a + a`. On success, return a struct with `.valid = true`, `variable = a`, and `multiplier = 2`.
+// The method matches `n` for pattern: AddNode(a, a).
+AddNode::Multiplication AddNode::find_simple_addition_pattern(const Node* n, BasicType bt) {
+  if (n->Opcode() == Op_Add(bt) && n->in(1) == n->in(2)) {
+    return Multiplication{true, n->in(1), 2};
+  }
+
+  return Multiplication::make_invalid();
+}
+
+// Try to match `n = a << CON`. On success, return a struct with `.valid = true`, `variable = a`, and
+// `multiplier = 1 << CON`.
+// Match `n` for pattern: LShiftNode(a, CON).
+// Note that the power-of-2 multiplication optimization could potentially convert a MulNode to this pattern.
+AddNode::Multiplication AddNode::find_simple_lshift_pattern(const Node* n, BasicType bt) {
+  // Note that power-of-2 multiplication optimization could potentially convert a MulNode to this pattern
+  if (n->Opcode() == Op_LShift(bt) && n->in(2)->is_Con()) {
+    Node* con = n->in(2);
+    if (!con->is_top()) {
+      return Multiplication{true, n->in(1), java_shift_left(1, con->get_int(), bt)};
+    }
+  }
+
+  return Multiplication::make_invalid();
+}
+
+// Try to match `n = CON * a`. On success, return a struct with `.valid = true`, `variable = a`, and `multiplier = CON`.
+// Match `n` for patterns:
+//     - (1) MulNode(CON, a)
+//     - (2) MulNode(a, CON)
+AddNode::Multiplication AddNode::find_simple_multiplication_pattern(const Node* n, BasicType bt) {
+  // This optimization technically only produces MulNode(CON, a), but we might as match MulNode(a, CON), too.
+  if (n->Opcode() == Op_Mul(bt) && (n->in(1)->is_Con() || n->in(2)->is_Con())) {
+    // Pattern (1)
+    Node* con = n->in(1);
+    Node* base = n->in(2);
+
+    // Pattern (2)
+    if (!con->is_Con()) {
+      // swap ConNode to lhs for easier matching
+      swap(con, base);
+    }
+
+    if (!con->is_top()) {
+      return Multiplication{true, base, con->get_integer_as_long(bt)};
+    }
+  }
+
+  return Multiplication::make_invalid();
+}
+
+// Try to match `n = (a << CON1) + (a << CON2)`. On success, return a struct with `.valid = true`, `variable = a`, and
+// `multiplier = (1 << CON1) + (1 << CON2)`.
+// Match `n` for patterns:
+//     - (1) AddNode(LShiftNode(a, CON), LShiftNode(a, CON))
+//     - (2) AddNode(LShiftNode(a, CON), a)
+//     - (3) AddNode(a, LShiftNode(a, CON))
+//     - (4) AddNode(a, a)
```

Comment:
You could drop the `Node` part from the cases here, to make it a bit more concise.

Reply:
Alternatively, you could do it with the `<<` operator like you did in `find_serial_addition_patterns`. I think that would be more consistent.